### PR TITLE
Increase sorting scalability via CytoTable metadata columns

### LINKED ISSUE #175: Explore value sorting determinism (and possible changes)
From #174:

> I found that `duckdb==0.10.1` entails lowered determinism for joined data result value sorting. We previously have relied on PyArrow to consistently sort all columns and their values for testing - this appears to also be potentially in need of adjustment. As a result, I've added SQL [`ORDER BY ALL`](https://duckdb.org/docs/sql/query_syntax/orderby.html#order-by-all) (which orders all values in all columns from left to right) to one test which consistently failed with comparisons. As a to-do, we should explore why PyArrow is unable to perform the same level of data organization for testing and possibly raise an issue with that project if the results aren't in alignment with the design. Alternatively, moving to `ORDER BY ALL` or some parallel may be necessary to address the existing PyArrow-based data sorting for testing (otherwise we may sporadically see issues over time).

Because of the importance of this issue, adding that we need example cases where the fix has been validated with larger than testing datasets.

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR seeks to refine #175 by increasing the performance through generated CytoTable metadata columns which are primarily beneficial during large join operations. Anecdotally, I noticed that `ORDER BY ALL` memory consumption for joined tables becomes very high when working with a larger dataset. Before this change, large join operations attempt to sort by all columns included in the join. After this change, only CytoTable metadata columns are used for sorting, decreasing the amount of processing required to create deterministic datasets.

I hope to further refine this work through #193 and #176, which would I feel provide additional insights concerning performance and best practice recommendations. I can also see how these might be required to validate things here, but didn't want to hold review comments (as these also might further inform efforts within those issues).

Closes #175 

## What is the nature of your change?

- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

**Review by gwaybio — COMMENTED** [review]
This is a big PR @d33bs - tough for me (with my limited expertise) to go through! @falquaddoomi can you give a look when you get a chance.

In general, I have one concern. If we end up deciding (or duckdb fixes this) to go back to the old way (where we had thought but didn't end up having potential issues with sorting), it seems this will be super difficult to disentangle. Or maybe I'm not thinking about this correctly? Could it be that this increasing scalability is independent of the previous solution which reduced speed?

**Comment by gwaybio** [comment]
(some additional context @falquaddoomi - we are needing to solve this for an upcoming project that will use cytotable heavily. Thanks!)

File: cytotable/convert.py [hunk 1]
```diff
@@ -698,15 +725,30 @@ def _prepare_join_sql(
...
+        [
+            f"{table}.{meta_column}"
+            for table in order_by_tables
+            for meta_column in CYOTABLE_META_COLUMN_TYPES
+        ]
+    )
 
-    return joins
+    # add the order by statements to the join
+    return joins + order_by_sql
```
Comment (falquaddoomi) [inline cytotable/convert.py:751]: I wonder if you could address @gwaybio's comment about disabling the new meta column ordering behavior by making this "order by" clause togglable via an argument to this method?

File: cytotable/convert.py [hunk 2]
```diff
@@ -317,9 +319,28 @@ def _source_chunk_to_parquet(
...
+                ' AS "cytotable_meta_source_path"'
+            ),
+            f"CAST( {offset} AS {CYOTABLE_META_COLUMN_TYPES['cytotable_meta_offset']}) AS \"cytotable_meta_offset\"",
+            (
+                f"CAST( (row_number() OVER ()) AS {CYOTABLE_META_COLUMN_TYPES['cytotable_meta_rownum']})"
+                ' AS "cytotable_meta_rownum"'
+            ),
+        ]
+        # add source table columns
+        + [
```
[the remark this answers is not in the document]
Comment (d33bs) [reply cytotable/convert.py:343]: Thanks, great point, I like the idea of a setting for this - will update accordingly.

File: cytotable/convert.py [hunk 3]
```diff
@@ -375,6 +396,7 @@ def _source_chunk_to_parquet(
                     table_name=str(source["table_name"]),
                     chunk_size=chunk_size,
                     offset=offset,
+                    add_cytotable_meta=True,
```
[the remark this answers is not in the document]
Comment (d33bs) [reply cytotable/convert.py:399]: Thanks, makes sense!

File: cytotable/convert.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (d33bs) [reply cytotable/convert.py:751]: Thanks - sounds like a good idea, I'll add this in.

**Comment by d33bs** [comment]
Thanks @gwaybio and @falquaddoomi for the reviews! I like the idea of an optional setting for this sorting mechanism, with a possible backup method which doesn't leverage CytoTable metadata.

Generally, I still feel that sorting should be required to guarantee no data loss with `LIMIT and OFFSET` because this aligns with both DuckDB's docs and general SQL guidance. A hypothesis about what was allowing this to succeed in earlier work: DuckDB may have successfully retained all data with `LIMIT and OFFSET` queries through low system process and thread competition. The failing tests for `LIMIT and OFFSET` I believe nearly always dealt with multithreaded behavior in `moto`, meaning procedures may have been subject to system scheduler decisions about which tasks to delay vs execute (or perhaps there were system thread or memory leaks of some kind).

While we plan to remove `moto` as a dependency by addressing #198, it feels fuzzy yet to me whether these challenges are all the same. For example, it could be that `moto` triggered a coincidental [mutation test](https://en.wikipedia.org/wiki/Mutation_testing) with regard to DuckDB thread behavior (giving us further software visibility through a mutated test state). It could have also been a "perfect storm" through a bug in DuckDB `>0.10.x,<1.0.0` combined with `moto`'s behavior in tests. Then again, this could all just be my imagination, I'm not sure!

**Comment by d33bs** [comment]
Note: Initially failing tests for [4ffe9c1](https://github.com/cytomining/CytoTable/pull/204/commits/4ffe9c12ebc684bf4eb7a717a34750c3e49f57aa) appeared to have something to do with a Poetry dependency failure (maybe fixed through a deploy by the time of a 3rd re-run?). I don't think these are related to CytoTable code as they were at the layer of Poetry installations.

Errors were:
`AttributeError: '_CountedFileLock' object has no attribute 'thread_safe'` from `virtualenv` and `filelock` site-packages.

**Comment by d33bs** [comment]
Thanks again @gwaybio and @falquaddoomi ! I've added some updates which make sorting optional through the use of parameters called `sort_output`. These changes retain the ability to keep output sorted and also an option to avoid it altogether (reverting to earlier CytoTable behavior). I've kept the default to `sort_output=True` as I feel this is the safest option for the time being, but understand there may be reasons to avoid it based on the data or performance desired.

**Review by falquaddoomi — APPROVED** [review]
Nice to see that you added the sort option! I anticipate somewhere down the line we could use that option as a means to compare the performance and correctness of sorting vs. not sorting.

**Comment by d33bs** [comment]
Cheers, thanks @falquaddoomi ! Agreed on comparisons; it will be interesting to see the contrast, excited to learn more!

**Pull request merged**